casanovo/data/annotate_db.py

```diff
@@ -0,0 +1,131 @@
+    seq = re.sub(r"(.*)\[43\.[0-9]*\]", r"+43.006\1", seq)
+    seq = re.sub(r"(.*)\[\-17\.[0-9]*\]", r"-17.027\1", seq)
+    seq = re.sub(r"(.*)\[25\.[0-9]*\]", r"+43.006-17.027\1", seq)
+    return seq
+
+
+def annotate_mgf(peak_path: str, tide_path: str, output: Optional[str]):
+    """
+    Accepts a directory containing the results of a successful tide search, and an .mgf file containing MS/MS spectra.
+    The .mgf file is then annotated in the SEQ field with all of the candidate peptides for each spectrum, as well as their target/decoy status.
+    This annotated .mgf can be given directly to Casanovo-DB to perfrom a database search.
+
+        Parameters
+        ----------
+        tide_dir_path : str
+            Path to the directory containing the results of a successful tide search.
+        mgf_file : str
+            Path to the .mgf file containing MS/MS spectra.
+        output_file : str
+            Path to where the annotated .mgf will be written.
+
+    """
+    logger = logging.getLogger("casanovo")
+    # Get paths to tide search text files
+    tdf_path = os.path.join(tide_path, "tide-search.target.txt")
+    ddf_path = os.path.join(tide_path, "tide-search.decoy.txt")
+    try:
+        target_df = pd.read_csv(
+            tdf_path, sep="\t", usecols=["scan", "sequence", "target/decoy"]
+        )
+        decoy_df = pd.read_csv(
+            ddf_path, sep="\t", usecols=["scan", "sequence", "target/decoy"]
+        )
+    except FileNotFoundError as e:
+        logger.error(
+            "Could not find tide search results in the specified directory. "
+            "Please ensure that the directory contains the following files: "
+            "tide-search.target.txt and tide-search.decoy.txt"
+        )
+        raise e
+
+    logger.info("Successfully read tide search results from %s.", tide_path)
+
+    df = pd.concat([target_df, decoy_df])
+    scan_groups = df.groupby("scan")[["sequence", "target/decoy"]]
+
+    scan_map = {}
+
+    for scan, item in scan_groups:
+        td_group = item.groupby("target/decoy")["sequence"].apply(list)
+        if "target" in td_group.index:
+            target_candidate_list = list(
+                map(
+                    _normalize_mods,
+                    td_group["target"],
+                )
+            )
+        else:
+            target_candidate_list = []
+            logger.warn(f"No target peptides found for scan {scan}.")
+        if "decoy" in td_group.index:
+            decoy_candidate_list = list(
+                map(
+                    _normalize_mods,
+                    td_group["decoy"],
+                )
+            )
+            decoy_candidate_list = list(
+                map(lambda x: "decoy_" + str(x), decoy_candidate_list)
+            )
+        else:
+            decoy_candidate_list = []
+            logger.warn(f"No decoy peptides found for scan {scan}.")
+
+        scan_map[scan] = target_candidate_list + decoy_candidate_list
```

Comment:
If a scan has no targets or decoys it should be skipped. I think here you would get an empty SEQ= field which may cause problems?

Reply:
small "if" check added to prevent this.